File: src/main/java/com/amazonaws/sample/cdk/discourse/RedisNestedStack.java
package com.amazonaws.sample.cdk.discourse;

import software.amazon.awscdk.NestedStack;
import software.amazon.awscdk.services.ec2.*;
import software.amazon.awscdk.services.elasticache.CfnReplicationGroup;
import software.amazon.awscdk.services.elasticache.CfnSubnetGroup;
import software.constructs.Construct;

import java.util.List;
import java.util.stream.Collectors;

public class RedisNestedStack extends NestedStack {
    private final Vpc vpc;
    private SecurityGroup securityGroup;
    private final SubnetConfiguration subnet;
    public CfnSubnetGroup redisSubnetGroup;
    public CfnReplicationGroup redis;

    public RedisNestedStack(final Construct scope, final String id, final Vpc vpc, final SecurityGroup securityGroup, final SubnetConfiguration subnet) {
        super(scope, id);
        this.vpc = vpc;
        this.securityGroup = securityGroup;
        this.subnet = subnet;
        createRedisSubnetGroup(id);
        createRedis(id);
    }

    public String getAttrPrimaryEndPointAddress() {
        return redis.getAttrPrimaryEndPointAddress();
    }

    private void createRedisSubnetGroup(final String id) {
        redisSubnetGroup = CfnSubnetGroup.Builder.create(this, id + "RedisSubnetGroup")
                .subnetIds(vpc.selectSubnets(SubnetSelection.builder().subnetGroupName(subnet.getName()).build()).getSubnets().stream().map(ISubnet::getSubnetId).collect(Collectors.toList()))
                .description("Redis subnet group")
                .build();
    }

    private void createRedis(final String id) {
        redis = CfnReplicationGroup.Builder.create(this, id + "Redis")
                .autoMinorVersionUpgrade(true)
                .engine("redis")
                .replicationGroupDescription("Discourse redis")
                .replicasPerNodeGroup(1)
                .numNodeGroups(1)
                .cacheNodeType("cache.t4g.medium")
                .multiAzEnabled(true)
                .securityGroupIds(List.of(securityGroup.getSecurityGroupId()))
                .build();
        redis.addDependsOn(redisSubnetGroup);
        redis.addPropertyOverride("CacheSubnetGroupName", redisSubnetGroup.getRef());
    }
}

File: lib/redis-nested-stack.ts
import {Construct} from 'constructs';
import {NestedStack, NestedStackProps} from "aws-cdk-lib";
import {SecurityGroup, SubnetConfiguration, Vpc} from "aws-cdk-lib/aws-ec2";
import {CfnReplicationGroup, CfnSubnetGroup} from "aws-cdk-lib/aws-elasticache";

interface RedisNestedStackProps extends NestedStackProps {
    readonly vpc: Vpc;
    readonly securityGroup: SecurityGroup;
    readonly subnet: SubnetConfiguration;
}

export class RedisNestedStack extends NestedStack {
    redis: CfnReplicationGroup;
    redisSubnetGroup: CfnSubnetGroup;

    constructor(scope: Construct, id: string, props: RedisNestedStackProps) {
        super(scope, id, props);
        this.redisSubnetGroup = new CfnSubnetGroup(this, id + "RedisSubnetGroup", {
            subnetIds: props.vpc.selectSubnets({
                subnetGroupName: props.subnet.name
            }).subnetIds,
            description: 'Redis subnet group'
        });

        this.redis = new CfnReplicationGroup(this, id + "Redis", {
            autoMinorVersionUpgrade: true,
            engine: 'redis',
            replicationGroupDescription: 'Discourse redis',
            replicasPerNodeGroup: 1,
            numNodeGroups: 1,
            cacheNodeType: 'cache.t4g.medium',
            multiAzEnabled: true,
            securityGroupIds: [props.securityGroup.securityGroupId]
        });
        this.redis.addDependsOn(this.redisSubnetGroup);
        this.redis.addPropertyOverride('CacheSubnetGroupName', this.redisSubnetGroup.ref);
    }
}
